flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -127,6 +127,10 @@ protected void cancelTask() {
 
 	@Override
 	protected void finishTask() throws Exception {
+		// We tell the mailbox to finish, to prevent any exceptions that might occur during
+		// finishing from leading to a FAILED state. This could happen, for example, when cancelling
+		// sources as part of a "stop-with-savepoint".
+		mailboxProcessor.allActionsCompleted();
```

Comment:
Sorry for being late in this PR.
Is there a reason why in `master` branch version there is `mailboxProcessor.allActionsCompleted();` in `finishTask()` but in `release-1.9` branch doesn't have it (#9115)?

Reply:
On the `release-1.9` branch `StreamTask.cancel()` also doesn't use this method, in fact the `MailboxProcessor` doesn't yet exist and isn't used like this. However, this solution leads to a race condition, see also https://github.com/apache/flink/pull/9125.